app/src/main/java/org/kiwix/kiwixmobile/webserver/KiwixServer.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2020 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.kiwix.kiwixmobile.webserver
+
+import android.util.Log
+import org.kiwix.kiwixlib.JNIKiwixException
+import org.kiwix.kiwixlib.JNIKiwixServer
+import org.kiwix.kiwixlib.Library
+import javax.inject.Inject
+
+private const val TAG = "KiwixServer"
+
+class KiwixServer @Inject constructor(private val jniKiwixServer: JNIKiwixServer) {
+  fun createKiwixServer(selectedBooksPath: ArrayList<String>): JNIKiwixServer? {
```

Comment:
this method should be a member of `KiwixServer.Factory` ie

```
class KiwixServer {
  class Factory{
     fun createKiwixServer(){ etc }
  }
}

Reply:
Did you find this advice confusing? I can explain it in more detail if you want to highlight the parts you are having trouble with

Reply:
Nope, this isn't confusing to me, I am able to do this. But to use this factory I have to create an instance of it in the serverHelper class and that is also cool, with this I'm able to create JNIKiwixServer also with the help of factory. But the problem occurs when I call `start` and `stop` server function, it uses the `jniKiwixServer` param which I pass in the constructor of the main class. 

And here comes the problem: it is unable to start the server(when I click concerned button in the device), most probably because we have two different jniKiwixServers being used(the one created from factory and the one that is being injected in the class).

But when I tried hacking it, I removed it as a param in the main class, now I declare this var jniKiwixServer as lateinit, and assign it inside the server creation function that lies inside the factory, with this being done, now I believe that the server being created, started and stopped are same. And when I run it in the device, application crashes saying `propery declared lateinit has not been intialized`, despite the fact I'm initializing it inside the facotry function.

Reply:
I am having an tough time parsing this, here is what the code should look like

```
class KiwixServer(private val jniKiwixServer:KiwixServer) {
  class Factory{
     fun createKiwixServer(selectedBooksPath: ArrayList<String>):KiwixServer{
        ...
       return KiwixServer(JNIKiwixServer(library))
   }
  }

 fun start(port:Int):Boolean{
    jniKiwixServer.setPort(port)
    return jniKiwixServer.start()
  }

//also a stop method
}
```

Reply:
One thing I would add that I'm writing start and stop server functions inside the factory class.

Reply:
A factory should only create. In the example above `KiwixServer.Factory` is a different class to `KiwixServer`, `KiwixServer` holds the responsibility of controlling jniKiwixServer. Here is the docs on [nested/inner classes](https://kotlinlang.org/docs/reference/nested-classes.html) they are a bit different from java's in their default behaviour

Reply:
Holy Jesus!! (for https://github.com/kiwix/kiwix-android/pull/2536#discussion_r542248089)